docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+The tenant attachment generation and node generation are mostly independent, but have one key
+coupling:
+
+- Control plane must not "steal" a tenant from a node (by attaching to another node without detaching
+  from the first node) without also incrementing the original node's generation number, which
+  effectively marks the original node as dead.
+- This enables nodes to trust that if their generation ID is still current, then all their
+  attachments are as well.
+
+### Object Key Changes
+
+All object keys (layer objects and index objects) will have a suffix as follows:
+
+```
+   <node id>-<attachment generation>-<node generation>
+```
+
+To avoid unreasonably long keys, we may make some reasonable assumptions about max lengths
+of these integers:
+
+- 32 bit attachment generation (this easily accommodates reasonable worst case scenarios like
+  a node restarting every minute for 5 years)
+- 16 bit node ID (this doesn't have to accommodate all pageservers ever, just all the pageservers
+  that act together in a logical cluster)
+- 32 bit node generation (as with attachment generation, this can handle even very frequent restarts)
+
+So an example suffix for generation 1 of node 0 might look like:
+
+```
+   00000001-0000-00000001
+```
+
+This suffix provides a lot of flexibility in how we may handle migration/restarts/failover
+at runtime:
+
+- We can replace a pageserver with another physical node having the same pageserver ID, and
+  the pageserver ID in the name will protect from conflicting writes.
+- Two pageservers may have the same tenant attached without risking stepping on each others
+  writes.
+- Ultimately, pageservers might self-coordinate such that a single "attachment" refers to
+  multiple pageservers, and they may internally do HA without troubling the control
+  plane to do new attachments, as the node generation makes this safe.
+
+### Deletion Changes
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. none of its attached tenants have been re-attached to another node).
```

Comment:
This is an interesting point.  We could do it either way.  Couple of thoughts:
- There's one other advantage to using node generation as a "nothing was stolen" flag that I didn't really call out here: it simplifies deletion code.  If we can handle deletions in a global timeline-agnostic queue (a la https://github.com/neondatabase/neon/pull/4960): if we use individual tenant attachment generations to gate deletion, then the overall deletion process needs a few more hooks from the timelines into the general deletion queue so that they can all individually gate the queue's progress on their generations being fresh.
- Aside from generation numbers, doing graceful transfers can be more efficient in I/O terms than a "hard" transfer, because it gives the prior pageserver an opportunity to flush to S3 before the new pageserver goes live.  That minimizes the window of recent LSNs that will be unreadable on the new pageserver until it ingests the WAL to catch up with the state that the prior pageserver might have only had on its on-disk content.

I'll think about it a bit more.

Reply:
Thoughts about the details a bit more overnight.

Deletions: I think the impact on deletion queue code is reasonable: it just has to keep a map of which tenants have pending deletions, and when it periodically queries the control plane for permission to proceed, it can bundle up all those attachement generations and the node generation into a single CP request.

Graceful transfers: I started defining the live migration process for the various permutations of a happy node, an unresponsive node, etc.  I think we can boil this down into one migration procedure that has an optional time-bounded "graceful" step where we call out to the old pageserver and advise it to go into a non-deleting mode and flush if possible.  If that graceful step times out, we can still go ahead with the attachment to a new pageserver and the old pageserver will find out via its periodic generation number check that it should go into that same non-deleting mode.

So it won't be precisely true that a clean migration is the same as a dirty migration (there's an extra step in a clean migration to reduce I/O waste), but that the extra step in the clean migration will be advisory and not necessary for correctness.

I'm revising this RFC to remove the requirement for hard-fencing a node when we attach a tenant elsewhere without advising the old node.  The fencing will now only happen if the control plane decides a node is dead (for HA purposes) or if two nodes with the same node ID get started.

Reply:
896bfdb2299d2a4f4b1b942a35f5704ce801d020